### ISSUE TITLE
CommentsIndentationCheck reports some false positives when in the middle of a method

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#2077]
Code:
````
package test;

public class TestClass3 {
	public void test() {
		int i = 0;
		String s = "";

		switch (i) {
		case -2:
			// what
			i++;
			// no break here
		case 0:
			// what
			s.indexOf("ignore");
			// no break here <-- line 16
		case 1:
		case 2:
			i--;
			// no break here <-- line 20
		case 3:
		}

		String breaks = ""
		// start <-- line 25
				+ "</table>"
				// middle <-- line 27
				+ ""
		// end <-- line 29
		;
	}
}
````
Config:
````
<module name="Checker">
	<property name="charset" value="ASCII" />
	<property name="severity" value="error" />
	<property name="fileExtensions" value="java" />
	
	<module name="TreeWalker">
		<module name="CommentsIndentation">
		</module>
	</module>
</module>
````
Output:
````
Starting audit...
TestClass3.java:16: Comment has incorrect indentation level 3, expected is 2, indentation should be the same level as line 17.
TestClass3.java:25: Comment has incorrect indentation level 2, expected is 4, indentation should be the same level as line 26.
Audit done.
Checkstyle ends with 2 errors.
````

The code is formatted with the default Eclipse formatter, so all errors are reported incorrectly.
From the source, it seems CommentsIndentationCheck is built to check the next statement while all the correct formattings seem to be more based on the line before it, but I can't specifically come up with a rule that includes line 27 and line 29. 

Line 16 was particularly hard to reproduce because of its issue with the "DOT" token type on the previous statement. If you notice at line 20, it follows the same pattern as 16, but no error is produced.

**Comment by romani** [comment]
>TestClass3.java:25

As for me that violation is correct. Please explain why you think it is false positive.

**Comment by romani** [comment]
@MEZk , please take a look.

**Comment by rnveach** [comment]
@romani 
Like I said, I used the default Eclipse (Kepler SR2) formatter. It put the comment there, so I assume it to place things correctly. If I move the comment to the same indentation as line 26, eclipse just moves it back when I run the formatter.

**Comment by MEZk** [comment]
@romani 
**Line 16**
See your comments here
https://github.com/checkstyle/checkstyle/pull/808#issuecomment-83128375

**Line 25,27,29**
I'll take a look

**Comment by romani** [comment]
> See your comments here

do you have question or it is just for my information ?

**Comment by MEZk** [comment]
@romani 
>do you have question or it is just for my information ?

It's information for you, the check was designed to compare indentation between comment and previous statement in thgose cases. But @rnveach was right, now we have the problem that when we try to reseive previous statement fromthe following example:

````java
case 0:
    // what
    s.indexOf("ignore"); // previous statement
    // no break here <-- target comment
````

it returns **DOT**-token, not **s**-token which begins the previous statement.

**Comment by MEZk** [comment]
@rnveach 
As concerns line 25.

Comments in general describe the next statement or code block. Exception is 'fall throught' comment in case block.

````java
case "5":
   foo();
   // fall through has indentation as previous line
case "6":
````

Comment on line 25 should describe the statement on line 26 and should have the same indentation as line 26. 

````java
String breaks = ""
// start <-- line 25
       + "</table>"
       // middle <-- line 27
       + ""
// end <-- line 29
;
````

We can not control all settings of Eclipse or other IDEs. Please, try to set up formatter properly, like I've done in IntelliJ IDEA.

Let's consider that we have the following example:

````java
System.out.println();  // line 1

// loop variable (line 3)
int i;  // line 4
````

Comment on line 3 describes the variable on line 4. It is a common practise to place comment before statement which it describes. And what about previous example? How can we determine whether user intends to describe the next statement or the previous one? Checkstyle can not control these cases.

**Comment by rnveach** [comment]
@MEZk 
I don't know other IDEs, so this could be IDE specific, but there doesn't seem to be a way to disable this in Eclipse.
I played with all the Formatting values in Eclipse Luna, nothing under the "Comments" will specifically make code and comments the same column position for consecutive lines. The closest option is "Line Wrapping" (specifically "Binary expressions" for my example) but this just controls the code's indentation, nothing will control the comment's.

Also, the documentation that this check is based on says "surrounding code", not "always the next line". (http://checkstyle.sourceforge.net/reports/google-java-style.html#s4.8.6.1-block-comment-style) Was this not worded like this for a reason? The comment does match the indentation of the previous line, which 'surrounds' the comment. I don't see the "no break" comment as an exception, but valid and still following the wording of the rule.

My intention for this comment was always the next line, like you said. In some instances, I used this to describe the group of lines below it, since that's what it looks like with the indentation that is made. I never use this style for the previous line, and would move it around to correct it if the formatter tried to do otherwise.

**Comment by romani** [comment]
@MEZk 

> We can not control all settings of Eclipse or other IDEs. Please, try to set up formatter properly if possible.

please prove that is possible by IntelijIdea

**Comment by romani** [comment]
@rnveach 

> Was this not worded like this for a reason?

Wording in Google style guide is known problem already, we have long discussion with Google to fix uncertain text,  Google team agree and do nothing :) .  All we could presume it is looking at their opensource code, that if full of violations for formatting. Each team in Google use its own style.

> The comment does match the indentation of the previous line, which 'surrounds' the comment.

And we have Checkstyle's limitations to clearly detect user intention of explanation target - above or below . We decided to go the most common way - comment describe line that is below. All other interesting formatting that user need - should go to suppression/exceptions.

> I never use this style for the previous line, and would move it around to correct it if the formatter tried to do otherwise.

We need to file issue against Eclipse, Eclipse always has problems with formatting (formatting between different versions is different, Mars release have a lot or regressions for formatting to wrapped code). Please do not hesitate to open issue against Eclipse.

**Comment by MEZk** [comment]
@rnveach, @romani 

For IntelliJ IDEA:

1) Go to File -> Settings -> Editor -> Code & Style -> Java -> Tabs and Intends and set the parameters of the editor as the following:

![screenshot](https://cloud.githubusercontent.com/assets/7242568/10330932/f8c8fb42-6cd9-11e5-8690-302e8bfd0f80.png)

2) Select the code you want to format and press Ctrl + Alt + I or go to Code -> Auto Indent Lines.

Before formatting, when I press 'enter' after line 24 it formats the code like this:
```java
String breaks = ""
// start <-- line 25
    + "</table>"
    // middle <-- line 27
    + ""
    // end <-- line 29
;
````

And after formatting:
````java
String breaks = ""
    // start <-- line 25
    + "</table>"
    // middle <-- line 27
    + ""
    // end <-- line 29
;
````

**Comment by MEZk** [comment]
@rnveach, @romani,
For NetBeans:

1) Go to Tools -> Options -> Editor -> Formatting. 
2) Chose Java language. 
3) Chose category 'Tabs and Intends'
4) Set up indentation parameters as in the following example:

![screenshot](https://cloud.githubusercontent.com/assets/7242568/10361846/82aa28d4-6db3-11e5-89aa-e3ed596feed1.png)

5) Select the code you want to format and press Alt + Shift + F or go to Source -> Format.
Before formatting, when I press 'enter' after line 24 it formats the code like this:
````java
String breaks = ""
// start <-- line 25
    + "</table>"
    // middle <-- line 27
    + ""
    // end <-- line 29
;
````
And after formatting:
````java
String breaks = ""
    // start <-- line 25
    + "</table>"
    // middle <-- line 27
    + ""
    // end <-- line 29
;
````

**Comment by romani** [comment]
@MEZk , good job.

@rnveach , now you have evidences that Eclipse is not corect, please open issue on Eclipse and put a link to it here, for future reference.

**Comment by romani** [comment]
fixed by #2286 .
